crates/sqlexec/src/remote/location_aware.rs

```diff
@@ -0,0 +1,109 @@
+use crate::planner::physical_plan::client_recv::ClientExchangeRecvExec;
+use async_trait::async_trait;
+use datafusion::error::Result;
+use datafusion::{
+    arrow::datatypes::SchemaRef,
+    datasource::TableProvider,
+    execution::context::SessionState,
+    logical_expr::{LogicalPlan, TableProviderFilterPushDown, TableType},
+    physical_plan::{ExecutionPlan, Statistics},
+    prelude::Expr,
+};
+use std::{any::Any, sync::Arc};
+use uuid::Uuid;
+
+/// Who should be scanning the table.
+#[derive(Debug, Clone, Copy)]
+pub enum ScanLocation {
+    /// Client should scan the table (e.g. local CSV files or data frames).
+    ///
+    /// This will require a data exchange.
+    Client {
+        /// ID to use when constructing the recv exec.
+        ///
+        /// When the server begins executing, it will attempt to pull from the
+        /// recv exec. The client will be responsible for sending record batches
+        /// to the server using this ID.
+        broadcast_id: Uuid,
+    },
+    /// Server should scan the table.
+    Server,
+}
+
+/// A table provider that is aware of where it should be scanned from.
+///
+/// Currently this is assuming that the server is executing this plan.
+pub struct LocationAwareTableProvider {
+    pub inner: Arc<dyn TableProvider>,
+    pub location: ScanLocation,
+}
+
+impl LocationAwareTableProvider {
+    pub fn new(inner: Arc<dyn TableProvider>, location: ScanLocation) -> Self {
+        Self { inner, location }
+    }
+}
+
+#[async_trait]
+impl TableProvider for LocationAwareTableProvider {
```

Comment:
Not yet. I wanted to avoid making a large change that started to touch other bits of the code. Ideally we'll be returning `LocationAwareTableProvider` instead of `Arc<dyn TableProvider>` from our table code, so that at a higher level we could do something like:

```rust
// Create associated send execs
let send_execs: Vec<_> = location_aware_tables
  .iter()
  .filter(|t| t.needs_client_scan())
  .map(|t| t.create_client_scan_send_exec())
  .collect();

// Do planning, serialize and send physical plans to remote here...

let mut handles = Vec::new();
for send_exec in send_execs {
  let stream = send_exec().execute(...)?;
  handles.push(tokio::spawn(async move {
    while let Some(result) = stream.next().await {
      let _ = result?; // The stream just returns number of rows sent.
    }
    Ok(())
  }));
}

// Join handles...
```